File: Assets/Scripts/Core/SpecialCardEffectRegistry.cs
using System;
using System.Collections.Generic;
using Bunker.Core.Effects;

namespace Bunker.Core
{
    // Maps a SpecialCard's string EffectType to its concrete implementation.
    // Using a string key (matching SpecialCardEffectType constants) instead of
    // an enum lets designers add new cards as data without a code change,
    // as long as a matching effect class is registered here.
    public class SpecialCardEffectRegistry
    {
        private readonly Dictionary<string, ISpecialCardEffect> _effects = new();

        public SpecialCardEffectRegistry()
        {
            Register(SpecialCardEffectType.RevealHiddenTrait, new RevealHiddenTraitEffect());
            Register(SpecialCardEffectType.VoteImmunity, new VoteImmunityEffect());
            Register(SpecialCardEffectType.SwapTrait, new SwapTraitEffect());
            Register(SpecialCardEffectType.ForceRevealAll, new ForceRevealAllEffect());
        }

        private void Register(string effectType, ISpecialCardEffect effect)
        {
            _effects[effectType] = effect;
        }

        public bool TryGet(string effectType, out ISpecialCardEffect effect)
        {
            return _effects.TryGetValue(effectType, out effect);
        }
    }
}

File: bunker-server/src/game/effects/SpecialCardEffectRegistry.ts
import { SpecialCardEffectType, SpecialCardEffectTypeValue } from "../SpecialCard";
import { ISpecialCardEffect } from "./ISpecialCardEffect";
import { RevealHiddenTraitEffect } from "./RevealHiddenTraitEffect";
import { VoteImmunityEffect } from "./VoteImmunityEffect";
import { SwapTraitEffect } from "./SwapTraitEffect";
import { ForceRevealAllEffect } from "./ForceRevealAllEffect";

// Maps a SpecialCard's effectType string to its concrete implementation.
// Same rationale as the C# registry: string-keyed rather than enum-keyed,
// so new cards can be added via content as long as a matching effect class
// is registered here.
export class SpecialCardEffectRegistry {
  private readonly effects = new Map<SpecialCardEffectTypeValue, ISpecialCardEffect>();

  constructor() {
    this.register(SpecialCardEffectType.RevealHiddenTrait, new RevealHiddenTraitEffect());
    this.register(SpecialCardEffectType.VoteImmunity, new VoteImmunityEffect());
    this.register(SpecialCardEffectType.SwapTrait, new SwapTraitEffect());
    this.register(SpecialCardEffectType.ForceRevealAll, new ForceRevealAllEffect());
  }

  private register(effectType: SpecialCardEffectTypeValue, effect: ISpecialCardEffect): void {
    this.effects.set(effectType, effect);
  }

  tryGet(effectType: SpecialCardEffectTypeValue): ISpecialCardEffect | undefined {
    return this.effects.get(effectType);
  }
}
